x-pack/plugins/enterprise_search/public/applications/app_search/components/schema/schema_logic.ts

```diff
@@ -0,0 +1,170 @@
+
+import { ADD_SCHEMA_ERROR, ADD_SCHEMA_SUCCESS, UPDATE_SCHEMA_SUCCESS } from './constants';
+import { SchemaBaseLogic, SchemaBaseValues, SchemaBaseActions } from './schema_base_logic';
+import { SchemaApiResponse } from './types';
+
+interface SchemaValues extends SchemaBaseValues {
+  isUpdating: boolean;
+  hasSchema: boolean;
+  hasSchemaChanged: boolean;
+  cachedSchema: Schema;
+  mostRecentIndexJob: Partial<IndexJob>;
+  unconfirmedFields: string[];
+  hasUnconfirmedFields: boolean;
+  hasNewUnsearchedFields: boolean;
+  isModalOpen: boolean;
+}
+
+interface SchemaActions extends SchemaBaseActions {
+  loadIndexedEngineSchema(): void;
+  onSchemaLoad(response: SchemaApiResponse): SchemaApiResponse;
+  addSchemaField(
+    fieldName: string,
+    fieldType: SchemaType
+  ): { fieldName: string; fieldType: SchemaType };
+  updateSchemaFieldType(
+    fieldName: string,
+    fieldType: SchemaType
+  ): { fieldName: string; fieldType: SchemaType };
+  updateSchema(successMessage?: string): string | undefined;
+  onSchemaUpdateError(): void;
+  openModal(): void;
+  closeModal(): void;
+}
+
+export const SchemaLogic = kea<MakeLogicType<SchemaValues, SchemaActions>>({
+  path: ['enterprise_search', 'app_search', 'schema_logic'],
+  connect: {
+    values: [SchemaBaseLogic, ['dataLoading', 'schema']],
+    actions: [SchemaBaseLogic, ['loadSchema', 'setSchema']],
+  },
+  actions: {
+    loadIndexedEngineSchema: true,
+    onSchemaLoad: (response) => response,
+    addSchemaField: (fieldName, fieldType) => ({ fieldName, fieldType }),
+    updateSchemaFieldType: (fieldName, fieldType) => ({ fieldName, fieldType }),
+    updateSchema: (successMessage) => successMessage,
+    onSchemaUpdateError: true,
+    openModal: true,
+    closeModal: true,
+  },
+  reducers: {
+    isUpdating: [
+      false,
+      {
+        updateSchema: () => true,
+        onSchemaLoad: () => false,
+        onSchemaUpdateError: () => false,
+      },
+    ],
+    cachedSchema: [
+      {},
+      {
+        onSchemaLoad: (_, { schema }) => schema,
+      },
+    ],
+    mostRecentIndexJob: [
+      {},
+      {
+        onSchemaLoad: (_, { mostRecentIndexJob }) => mostRecentIndexJob,
+      },
+    ],
+    unconfirmedFields: [
+      [],
+      {
+        onSchemaLoad: (_, { unconfirmedFields }) => unconfirmedFields,
+      },
+    ],
+    hasNewUnsearchedFields: [
+      false,
+      {
+        onSchemaLoad: (_, { unsearchedUnconfirmedFields }) => unsearchedUnconfirmedFields,
+      },
+    ],
+    isModalOpen: [
+      false,
+      {
+        openModal: () => true,
+        closeModal: () => false,
+        onSchemaLoad: () => false,
+        onSchemaUpdateError: () => false,
+      },
+    ],
+  },
+  selectors: {
+    hasSchema: [(selectors) => [selectors.schema], (schema) => Object.keys(schema).length > 0],
+    hasSchemaChanged: [
+      (selectors) => [selectors.schema, selectors.cachedSchema],
+      (schema, cachedSchema) => !isEqual(schema, cachedSchema),
```

Comment:
Not really sure if there's a more performant way to check for whether in-state schema has changed vs cachedSchema, feel free to shout if you have ideas. I just copied the existing ent-search logic for now